Modules/_remote_debugging/binary_io_writer.c

```diff
@@ -810,22 +839,49 @@ build_frame_stack(BinaryWriter *writer, PyObject *frame_list,
         /* Use unchecked accessors since we control the data structures */
         PyObject *frame_info = PyList_GET_ITEM(frame_list, k);
 
-        /* Get filename, location, funcname from FrameInfo using unchecked access */
+        /* Get filename, location, funcname, opcode from FrameInfo using unchecked access */
         PyObject *filename = PyStructSequence_GET_ITEM(frame_info, 0);
         PyObject *location = PyStructSequence_GET_ITEM(frame_info, 1);
         PyObject *funcname = PyStructSequence_GET_ITEM(frame_info, 2);
+        PyObject *opcode_obj = PyStructSequence_GET_ITEM(frame_info, 3);
+
+        /* Extract location fields (can be None for synthetic frames) */
+        int32_t lineno = LOCATION_NOT_AVAILABLE;
+        int32_t end_lineno = LOCATION_NOT_AVAILABLE;
+        int32_t column = LOCATION_NOT_AVAILABLE;
+        int32_t end_column = LOCATION_NOT_AVAILABLE;
 
-        /* Extract lineno from location (can be None for synthetic frames) */
-        int32_t lineno = 0;
         if (location != Py_None) {
-            /* Use unchecked access - first element is lineno */
+            /* LocationInfo is a struct sequence or tuple with:
+             * (lineno, end_lineno, col_offset, end_col_offset) */
             PyObject *lineno_obj = PyTuple_Check(location) ?
                 PyTuple_GET_ITEM(location, 0) :
                 PyStructSequence_GET_ITEM(location, 0);
-            lineno = (int32_t)PyLong_AsLong(lineno_obj);
+            PyObject *end_lineno_obj = PyTuple_Check(location) ?
+                PyTuple_GET_ITEM(location, 1) :
+                PyStructSequence_GET_ITEM(location, 1);
+            PyObject *column_obj = PyTuple_Check(location) ?
+                PyTuple_GET_ITEM(location, 2) :
+                PyStructSequence_GET_ITEM(location, 2);
+            PyObject *end_column_obj = PyTuple_Check(location) ?
+                PyTuple_GET_ITEM(location, 3) :
+                PyStructSequence_GET_ITEM(location, 3);
+
+            PYLONG_TO_INT32_OR_DEFAULT(lineno_obj, lineno, LOCATION_NOT_AVAILABLE);
+            PYLONG_TO_INT32_OR_DEFAULT(end_lineno_obj, end_lineno, LOCATION_NOT_AVAILABLE);
+            PYLONG_TO_INT32_OR_DEFAULT(column_obj, column, LOCATION_NOT_AVAILABLE);
+            PYLONG_TO_INT32_OR_DEFAULT(end_column_obj, end_column, LOCATION_NOT_AVAILABLE);
+        }
+
+        /* Extract opcode (can be None) */
+        uint8_t opcode = OPCODE_NONE;
+        if (opcode_obj != Py_None) {
+            long opcode_long = PyLong_AsLong(opcode_obj);
             if (UNLIKELY(PyErr_Occurred() != NULL)) {
                 PyErr_Clear();
-                lineno = 0;
+                opcode = OPCODE_NONE;
+            } else if (opcode_long >= 0 && opcode_long <= 254) {
```

Comment:
This could use `MAX_REAL_OPCODE`:

```suggestion
            } else if (opcode_long >= 0 && opcode_long <= MAX_REAL_OPCODE) {
```